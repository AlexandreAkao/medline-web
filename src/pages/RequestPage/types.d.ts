type RequestFormTypes = 'description' | 'text' | 'type' | 'file';

interface IUserRegisterActiona {
  type: RequestFormTypes;
  payload: string | number | File;
}
