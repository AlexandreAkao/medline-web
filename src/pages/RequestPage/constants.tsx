// [Receita, Atestado]
export const REQUEST_OPTIONS = [
  {
    5: 'Rivotril',
    6: 'Bacitracina',
    7: 'Metronidazol',
    8: 'Loxotan',
    9: 'Ritalina',
  },
  {
    0: 'Covid-19',
    1: 'Virose',
    2: 'Doen√ßa da mosca',
    3: 'Dengue',
    4: 'Gripe',
  },
];
