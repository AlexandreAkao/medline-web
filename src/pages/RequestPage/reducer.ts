export const requestFormReducer = (state: IRequestForm, action: IUserRegisterActiona) => {
  switch (action.type) {
    case 'description':
    case 'text':
    case 'type':
    case 'file':
      return { ...state, [action.type]: action.payload };
    default:
      return state;
  }
};

export const requestFormInitialState: IRequestForm = {
  description: -1,
  text: '',
  type: 1,
  file: null,
};
