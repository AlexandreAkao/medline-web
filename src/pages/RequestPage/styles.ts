import styled from 'styled-components';

import colors from 'styles/colors';
import { ButtonContainer } from 'components/Button/styles';

export const RequestContainer = styled.main`
  background-color: ${colors.primary.normal};
  height: 100%;
  overflow: auto;
`;

export const RequestForm = styled.form`
  display: flex;
  height: calc(100% - 80px);
  flex-direction: column;
  align-items: center;
`;

export const RequestFieldsContainer = styled.div`
  width: 100%;
  display: flex;
`;

export const RequestDescriptionContainer = styled.div`
  flex: 1;
  padding: 3rem;
  display: flex;
  flex-direction: column;
`;

export const RequestDescriptionTitle = styled.h2`
  color: ${colors.white.normal};
  margin-bottom: 4rem;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

export const RequestDescription = styled.textarea`
  height: 100px;
  border-radius: 8px;
  flex: 1;
  resize: none;
  padding: 16px;
  font-size: 1.125rem;
`;

export const RequestOptionContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 3rem;
  flex: 1;
  height: 100%;
`;

export const RequestFormTitle = styled.h2`
  color: ${colors.white.normal};
  margin-bottom: 4rem;
`;

export const RequestFormDescription = styled.div`
  background: ${colors.white.normal};
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  width: 100%;
  padding: 2rem;
  margin: 2rem 0;
`;

export const RequestFormDescriptionTitle = styled.h5`
  font-size: 1rem;
  font-weight: 500;
  margin-bottom: 2.5rem;
`;

export const RequestFormDescriptionCheckboxs = styled.div`
  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: center;
  gap: 1.125rem;
`;

export const RequestFormButtonContainer = styled.div`
  display: flex;
  justify-content: center;
  width: 100%;
  margin-top: 3rem;

  ${ButtonContainer} {
    display: flex;
    align-items: center;
    margin: 0 1rem;

    svg {
      margin-right: 0.5rem;
    }
  }
`;
