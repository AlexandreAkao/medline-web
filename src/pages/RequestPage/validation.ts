import { isEmpty, isNil } from 'ramda';

export const isFormValid = (form: IRequestForm) => {
  const values = Object.values(form);

  // eslint-disable-next-line no-restricted-syntax
  for (const value of values) {
    if (isEmpty(value) || isNil(value)) return false;
  }

  return true;
};
