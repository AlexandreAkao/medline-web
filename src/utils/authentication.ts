import { isEmpty } from 'ramda';

export const getUserInfo = () => {
  const user = JSON.parse(localStorage.getItem('user') ?? '{}');

  return isEmpty(user) ? null : (user as IUser);
};

export const geToken = () => localStorage.getItem('token');
