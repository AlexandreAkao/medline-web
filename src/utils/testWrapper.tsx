import { render } from '@testing-library/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (component: React.ReactNode) =>
  render(<MemoryRouter>{component}</MemoryRouter>);
