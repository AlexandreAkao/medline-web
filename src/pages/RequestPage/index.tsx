import { FormEvent, useReducer, useState } from 'react';
import { FaPaperclip, FaRegPaperPlane } from 'react-icons/fa';
import { toast } from 'react-toastify';
import { useNavigate } from 'react-router-dom';

import Button from 'components/Button';
import MedlineHeader from 'components/Header/MedlineHeader';
import Switch from 'components/Switch';
import {
  RequestContainer,
  RequestForm,
  RequestOptionContainer,
  RequestFormTitle,
  RequestFormDescription,
  RequestFormDescriptionTitle,
  RequestFormDescriptionCheckboxs,
  RequestFormButtonContainer,
  RequestDescriptionContainer,
  RequestDescription,
  RequestDescriptionTitle,
  RequestFieldsContainer,
} from 'pages/RequestPage/styles';
import FileInput from 'components/FileInput';
import Radio from 'components/Radio';
import { requestFormInitialState, requestFormReducer } from 'pages/RequestPage/reducer';
import { REQUEST_OPTIONS } from 'pages/RequestPage/constants';
import { isFormValid } from './validation';
import api from 'service/api';
import colors from 'styles/colors';
import Loader from 'components/Loader';

function RequestPage() {
  const [isLoading, setIsLoading] = useState(false);
  const navigate = useNavigate();

  const [requestForm, dispatchRequestForm] = useReducer(requestFormReducer, requestFormInitialState);

  const handleChangeForm = (payload: string | number | File, type: RequestFormTypes) => {
    dispatchRequestForm({ type, payload });
  };

  const handleSubmitRequest = async (event: FormEvent<HTMLFormElement>) => {
    try {
      setIsLoading(true);
      event.preventDefault();

      if (!isFormValid(requestForm)) {
        toast.error('Preencha todos os campos', { theme: 'colored' });
        return;
      }

      const { description, text, type, file } = requestForm;
      const formData = new FormData();
      formData.append('request', JSON.stringify({ description, text, type }));
      formData.append('file', file as File);

      const requestReponse = await api.post('request', formData);

      if (requestReponse.status === 200) {
        toast.success('Solicitação criada com Sucesso', { theme: 'colored' });
        navigate('/prescription');
      }
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <RequestContainer>
      {isLoading && <Loader />}
      <MedlineHeader />
      <RequestForm onSubmit={handleSubmitRequest}>
        <RequestFieldsContainer>
          <RequestDescriptionContainer>
            <RequestDescriptionTitle>Descrição</RequestDescriptionTitle>
            <RequestDescription
              placeholder="Descreva o seu problema"
              onChange={event => handleChangeForm(event.target.value, 'text')}
            />
          </RequestDescriptionContainer>
          <RequestOptionContainer>
            <RequestFormTitle>Solicitações</RequestFormTitle>
            <Switch
              optionLeft={{ label: 'Atestado', value: 1 }}
              optionRight={{ label: 'Receita', value: 0 }}
              onChange={event => {
                handleChangeForm(event.value, 'type');
                handleChangeForm(-1, 'description');
              }}
            />
            <RequestFormDescription>
              <RequestFormDescriptionTitle>Selecione uma das opções:</RequestFormDescriptionTitle>
              <RequestFormDescriptionCheckboxs>
                {Object.entries(REQUEST_OPTIONS[requestForm.type]).map(([key, value]) => (
                  <Radio
                    primary
                    key={key}
                    id={key}
                    name="option"
                    onChange={event => handleChangeForm(event.target.value, 'description')}
                    value={key}
                  >
                    {value}
                  </Radio>
                ))}
              </RequestFormDescriptionCheckboxs>
            </RequestFormDescription>
          </RequestOptionContainer>
        </RequestFieldsContainer>
        <RequestFormButtonContainer>
          <FileInput primary={false} onChange={files => handleChangeForm(files[0], 'file')}>
            <FaPaperclip color={requestForm.file ? colors.primary.normal : colors.black.normal} />
            Anexar
          </FileInput>
          <Button primary={false} type="submit" disabled={isLoading}>
            <FaRegPaperPlane />
            Enviar
          </Button>
        </RequestFormButtonContainer>
      </RequestForm>
    </RequestContainer>
  );
}

export default RequestPage;
