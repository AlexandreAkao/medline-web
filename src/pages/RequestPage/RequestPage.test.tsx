import { renderWithProvider } from 'utils/testWrapper';
import RequestPage from 'pages/RequestPage';

describe('RequestPage', () => {
  const renderComponent = () => {
    return renderWithProvider(<RequestPage />);
  };

  it('should match with snapshot', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
});
