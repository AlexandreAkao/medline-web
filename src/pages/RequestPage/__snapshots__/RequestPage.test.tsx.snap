// Vitest Snapshot v1

exports[`RequestPage > should match with snapshot 1`] = `
<div>
  <main
    class="sc-fnykZs bwVpRF"
  >
    <header
      class="sc-dkzDqf ifZngJ"
      data-testid="test-navigation"
    >
      <a
        href="/"
        id="header-logo"
      >
        <img
          alt="Medline logo"
          class="sc-hKMtZM kcTwZQ"
          src="/src/assets/images/medline-logo.svg"
        />
      </a>
      <nav
        class="sc-eCYdqJ dDvGMN"
        data-testid="test-navigation-items"
      >
        <div
          class="sc-gsnTZi eqeYZu"
        >
          <button
            class="sc-bczRLJ bJBahq"
            type="button"
          >
            Cadastro
          </button>
        </div>
        <div
          class="sc-gsnTZi eqeYZu"
        >
          <a
            href="/login"
          >
            Entrar
          </a>
        </div>
      </nav>
    </header>
    <form
      class="sc-fEOsli cWYXxm"
    >
      <div
        class="sc-bjUoiL bixoHL"
      >
        <div
          class="sc-idiyUo gUFoni"
        >
          <h2
            class="sc-dIouRR eAjdJH"
          >
            Descrição
          </h2>
          <textarea
            class="sc-hHLeRK haKSTc"
            placeholder="Descreva o seu problema"
          />
        </div>
        <div
          class="sc-dmRaPn jtdwDf"
        >
          <h2
            class="sc-kgflAQ gAmHCo"
          >
            Solicitações
          </h2>
          <div
            class="sc-crXcEl evBWJB"
          >
            <input
              class="sc-hAZoDl bWZkFr"
              data-testid="test-switch"
              id="react-switch-new"
              name="react-switch-new"
              type="checkbox"
            />
            <label
              class="sc-ksZaOG btGNZc"
              for="react-switch-new"
            >
              <div
                class="sc-evZas iEgbiY"
              >
                <span>
                  Atestado
                </span>
                <span>
                  Receita
                </span>
              </div>
              <span
                class="sc-breuTD izwwrN"
              />
            </label>
          </div>
          <div
            class="sc-fLlhyt jQodRf"
          >
            <h5
              class="sc-bBrHrO fibDkr"
            >
              Selecione uma das opções:
            </h5>
            <div
              class="sc-ivTmOn ewEhhN"
            >
              <div
                class="sc-ezWOiH eAiRkF"
              >
                <input
                  class="sc-iIPllB RThe"
                  id="0"
                  name="option"
                  type="radio"
                  value="0"
                />
                <label
                  class="sc-gicCDI hLFGNV"
                  for="0"
                >
                  Covid-19
                </label>
              </div>
              <div
                class="sc-ezWOiH eAiRkF"
              >
                <input
                  class="sc-iIPllB RThe"
                  id="1"
                  name="option"
                  type="radio"
                  value="1"
                />
                <label
                  class="sc-gicCDI hLFGNV"
                  for="1"
                >
                  Virose
                </label>
              </div>
              <div
                class="sc-ezWOiH eAiRkF"
              >
                <input
                  class="sc-iIPllB RThe"
                  id="2"
                  name="option"
                  type="radio"
                  value="2"
                />
                <label
                  class="sc-gicCDI hLFGNV"
                  for="2"
                >
                  Doença da mosca
                </label>
              </div>
              <div
                class="sc-ezWOiH eAiRkF"
              >
                <input
                  class="sc-iIPllB RThe"
                  id="3"
                  name="option"
                  type="radio"
                  value="3"
                />
                <label
                  class="sc-gicCDI hLFGNV"
                  for="3"
                >
                  Dengue
                </label>
              </div>
              <div
                class="sc-ezWOiH eAiRkF"
              >
                <input
                  class="sc-iIPllB RThe"
                  id="4"
                  name="option"
                  type="radio"
                  value="4"
                />
                <label
                  class="sc-gicCDI hLFGNV"
                  for="4"
                >
                  Gripe
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="sc-cxabCf fkcPDf"
      >
        <button
          class="sc-bczRLJ GABYc"
          type="button"
        >
          <svg
            color="#000"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(0, 0, 0);"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M43.246 466.142c-58.43-60.289-57.341-157.511 1.386-217.581L254.392 34c44.316-45.332 116.351-45.336 160.671 0 43.89 44.894 43.943 117.329 0 162.276L232.214 383.128c-29.855 30.537-78.633 30.111-107.982-.998-28.275-29.97-27.368-77.473 1.452-106.953l143.743-146.835c6.182-6.314 16.312-6.422 22.626-.241l22.861 22.379c6.315 6.182 6.422 16.312.241 22.626L171.427 319.927c-4.932 5.045-5.236 13.428-.648 18.292 4.372 4.634 11.245 4.711 15.688.165l182.849-186.851c19.613-20.062 19.613-52.725-.011-72.798-19.189-19.627-49.957-19.637-69.154 0L90.39 293.295c-34.763 35.56-35.299 93.12-1.191 128.313 34.01 35.093 88.985 35.137 123.058.286l172.06-175.999c6.177-6.319 16.307-6.433 22.626-.256l22.877 22.364c6.319 6.177 6.434 16.307.256 22.626l-172.06 175.998c-59.576 60.938-155.943 60.216-214.77-.485z"
            />
          </svg>
          Anexar
        </button>
        <input
          class="sc-llJcti cCkUrQ"
          data-testid="file-input-test"
          type="file"
        />
        <button
          class="sc-bczRLJ GABYc"
          type="submit"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M440 6.5L24 246.4c-34.4 19.9-31.1 70.8 5.7 85.9L144 379.6V464c0 46.4 59.2 65.5 86.6 28.6l43.8-59.1 111.9 46.2c5.9 2.4 12.1 3.6 18.3 3.6 8.2 0 16.3-2.1 23.6-6.2 12.8-7.2 21.6-20 23.9-34.5l59.4-387.2c6.1-40.1-36.9-68.8-71.5-48.9zM192 464v-64.6l36.6 15.1L192 464zm212.6-28.7l-153.8-63.5L391 169.5c10.7-15.5-9.5-33.5-23.7-21.2L155.8 332.6 48 288 464 48l-59.4 387.3z"
            />
          </svg>
          Enviar
        </button>
      </div>
    </form>
  </main>
  ,
</div>
`;
